specs/dataclass_transforms.txt

```diff
@@ -0,0 +1,399 @@
+        order: bool = True,
+    ):
+        ...
+
+class ModelBase(metaclass=ModelMeta):
+    ...
+
+
+# Example of how this class would be used by code that imports
+# from this library:
+class Customer(ModelBase, init=False, frozen=True, eq=False, order=False):
+    id: int
+    name: str
+```
+
+
+
+Field descriptors
+-----------------
+
+Most libraries that support dataclass-like semantics provide one or more
+"field descriptor" types that allow a class definition to provide additional
+metadata about each field in the class. This metadata can describe, for example,
+default values or indicate whether the field should be included in the
+synthesized `__init__` method.
+
+Field descriptors can be omitted in cases where additional metadata is not
+required or the only metadata is a statically-defined default value.
+
+```python
+@dataclass
+class Employee:
+    # Field with no descriptor
+    name: str
+
+    # Field that uses field descriptor class instance
+    age: Optional[int] = field(default=None, init=False)
+
+    # Field with type annotation and simple initializer to
+    # describe default value
+    is_paid_hourly: bool = True
+
+    # Field with inferred type and simple initializer to
+    # describe default value
+    office_number = "unassigned"
+```
+
+Libraries that support dataclass-like semantics and support field descriptor
+classes typically use common parameter names to construct these field
+descriptors. This specification formalizes the names and meanings of the
+parameters that must be understood for static type checkers. These standardized
+parameters must be keyword-only parameters. Field descriptor classes are
+allowed to use other parameters in their constructors, and those parameters
+can be positional and may use other names.
+
+
+`init` is an optional bool parameter that indicates whether the field should
+be included in the synthesized `__init__` method. If unspecified, it defaults
+to True.
+
+`default` is an optional parameter that provides the default value for the
+field.
+
+`default_factory` is an optional parameter that provides a runtime callback
+that returns the default value for the field. If `default` and `default_value`
+are both unspecified, the field is assumed to have no default value and must be
+provided a value when the class is instantiated.
+
+`alias` is an optional str parameter that provides an alternative name for
+the field. This alternative name is used in the synthesized `__init__` method.
+
+
+This example demonstrates 
+```python
+# Library code (within type stub or inline):
+class ModelField:
+    def __init__(
+        self,
+        *,
+        default: Optional[Any] = ...,
+        init: Optional[bool] = True,
+        **kwargs: Any
+    ) -> None: ...
+
+@typing.dataclass_transform(kw_only_default=True, field_descriptors=(ModelField, ))
+def create_model(cls: type) -> Callable[[_T], _T]: ...
+
+
+# Code that imports this library:
+@create_model(init=False)
```

Comment:
so init:False has no meaning in dataclass_transform (init seems not part of dataclass_trasform's known parameters)? and ModelField has init == true as default, and init in field descriptor seems part of known parameters, so will it affect how type checker treat the type?

Reply:
This parameter instructs the transform function `create_model` to not synthesize an `__init__` method for the decorated class. The stdlib `dataclass` decorator supports the same semantics.

```python
@dataclass(init=False)
class Foo:
    a: int

Foo(a=3) # Error because Foo has no __init__ method
```

When the `init` parameter is used in a field descriptor class like `ModelField`, it has a slightly different meaning. In particular, it means that this field should not be included in the synthesized `__init__` method.

Reply:
I see. so this just follow whatever dataclass existing rule.